File: 2026-01-January-GFG-POTD/January 10 - Substrings with K Distinct/Solution.py
class Solution:
    def countSubstr(self, s: str, k: int) -> int:
        def atMostK(s, k):
            if k < 0:
                return 0
            freq = [0] * 26
            left = 0
            distinct = 0
            count = 0

            for right in range(len(s)):
                r = ord(s[right]) - ord('a')
                if freq[r] == 0:
                    distinct += 1
                freq[r] += 1

                while distinct > k:
                    l = ord(s[left]) - ord('a')
                    freq[l] -= 1
                    if freq[l] == 0:
                        distinct -= 1
                    left += 1

                count += right - left + 1
            return count

        return atMostK(s, k) - atMostK(s, k - 1)

File: 2026-01-January-GFG-POTD/January 10 - Substrings with K Distinct/Solution.java
class Solution {
        public int countSubstr(String s, int k) {
        if(k == 0) return 0;
        return (int)(atMostK(s, k) - atMostK(s, k - 1));
    }
    
    private long atMostK(String s, int k){
        int n = s.length();
        int[] freq = new int[26];
        int left = 0;
        long count = 0;
        int distinct = 0;
        
        for(int right = 0; right < n; right++){
            int r = s.charAt(right) - 'a';
            if(freq[r] == 0) distinct++;
            freq[r]++;
            
            while(distinct > k){
                int l = s.charAt(left) - 'a';
                freq[l]--;
                if(freq[l] == 0) distinct--;
                left++;
            }
            
            // Number of substrings ending at right
            count += right - left + 1;
        }
        
        return count;
    }

}
